worker/upgradedatabase/worker.go

```diff
@@ -226,28 +302,125 @@ func (w *upgradeDBWorker) watchUpgrade() error {
-	toVersion := w.cfg.ToVersion
-	if fromVersion == toVersion {
-		w.cfg.Logger.Infof("database upgrade for %v already completed", toVersion)
-		w.cfg.DBUpgradeCompleteLock.Unlock()
+	if w.fromVersion == w.toVersion {
+		w.logger.Infof("database upgrade for %v already completed", w.toVersion)
+		w.dbUpgradeCompleteLock.Unlock()
 		return true
 	}
 
 	return false
 }
 
-func (w *upgradeDBWorker) runUpgrade(modelUUID domainupgrade.UUID) error {
-	w.cfg.Logger.Infof("running database upgrade database")
+func (w *upgradeDBWorker) runUpgrade(upgradeUUID domainupgrade.UUID) error {
+	w.logger.Infof("running database upgrade from: %v to: %v", w.fromVersion, w.toVersion)
+
+	ctx, cancel := w.scopedContext()
+	defer cancel()
+
+	if err := w.upgradeService.SetControllerReady(ctx, upgradeUUID, w.controllerID); err != nil {
+		return errors.Annotatef(err, "unable to set controller ready")
+	}
 
-	// This dummy worker, so we can just Unlock...
-	w.cfg.Logger.Infof("database upgrade already completed")
-	w.cfg.DBUpgradeCompleteLock.Unlock()
+	// Watch for the upgrade to be ready. This should ensure that all
+	// controllers are sync'd and waiting for the leader to start the upgrade.
+	watcher, err := w.upgradeService.WatchForUpgradeReady(ctx, upgradeUUID)
+	if err != nil {
+		return errors.Trace(err)
+	}
 
+	for {
+		select {
+		case <-w.catacomb.Dying():
+			return w.catacomb.ErrDying()
+
+		case <-w.clock.After(defaultUpgradeTimeout):
+			if err := w.upgradeService.SetDBUpgradeFailed(ctx, upgradeUUID); err != nil {
+				return errors.Annotatef(err, "unable to set db upgrade failed")
+			}
+
+			return errors.Errorf("timed out waiting for upgrade from: %v to: %v", w.fromVersion, w.toVersion)
+
+		case <-watcher.Changes():
+			if err := w.upgradeService.StartUpgrade(ctx, upgradeUUID); err != nil {
+				return errors.Annotatef(err, "unable to start upgrade")
+			}
+
+			// Upgrade the controller database first.
+			if err := w.upgradeController(ctx); err != nil {
+				return errors.Trace(err)
+			}
+			// Then upgrade the models databases.
+			if err := w.upgradeModels(ctx); err != nil {
+				return errors.Trace(err)
+			}
+
+			if err := w.upgradeService.SetDBUpgradeCompleted(ctx, upgradeUUID); err != nil {
+				return errors.Annotatef(err, "unable to set db upgrade completed")
+			}
+
+			w.logger.Infof("database upgrade already completed")
+			w.dbUpgradeCompleteLock.Unlock()
+
+			return nil
+		}
+	}
+}
+
+func (w *upgradeDBWorker) upgradeController(ctx context.Context) error {
+	w.logger.Infof("upgrading controller database from: %v to: %v", w.fromVersion, w.toVersion)
+
+	nodeID, _, err := w.controllerNodeService.DqliteNode(ctx, w.controllerID)
+	if err != nil {
+		return errors.Annotatef(err, "getting dqlite node id")
+	}
+
+	db, err := w.dbGetter.GetDB(coredatabase.ControllerNS)
+	if err != nil {
+		return errors.Annotatef(err, "controller db")
+	}
+
+	schema := schema.ControllerDDL(nodeID)
+	changeSet, err := schema.Ensure(ctx, db)
```

Comment:
Oh, actually, this will fail if the the node running the upgrade is not the bootstrapped node.

The patch is affected by the argument, which in turn affects the hash. They won't match.